Add FileWatcher to load sarif logs from .sarif folder automatically

yongyan-gh opened the pull request:
Fixes #380

* Automatically load all sarif log files from .sarif folder under solution directory to error lists when solution is loaded.
* Close all sarif logs when solution is closed.
* Monitor the folder and watch sarif log files created/deleted, load/unload results accordingly.
* Updates error list when any sarif log file in the folder is updated.
* Add option for enabling/disabling the behavior using option.
* Unit tests.

eddynaka commented on src/Sarif.Viewer.VisualStudio/Constants.cs:
```diff
@@ -29,5 +29,7 @@ public static class FileAndForgetFaultEventNames
             public const string TelemetryWriteEvent = "Microsoft/SARIF/Viewer/Telemetry/WriteEvent/Failed";
         }
+
+        public const string SarifFileSearchPattern = "*.sarif";
```
>SarifFileSearchPattern [](start = 28, length = 22)

move this near line 8

yongyan-gh replied:
fixed

---
In reply to: [597725678](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r597725678) [](ancestors = 597725678)

eddynaka commented on src/Sarif.Viewer.VisualStudio/Options/SarifViewerOptions.cs:
```diff
@@ -0,0 +1,80 @@
+            {
+                try
+                {
+                    return this.optionPage != null ? this.optionPage.MonitorSarifFolder : this.shouldMonitorSarifFolderDefaultValue;
```
>this.optionPage [](start = 27, length = 15)

I think you could change to:
return this.optionPage?.MonitorSarifFolder ?? this.shouldMonitorSarifFolderDefaultValue

eddynaka replied:
Also, i think you don't need the try/catch, unless you are expecting something else

---
In reply to: [597772222](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r597772222) [](ancestors = 597772222)

yongyan-gh replied:
fixed

---
In reply to: [597772640](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r597772640) [](ancestors = 597772640,597772222)

eddynaka commented on src/Sarif.Viewer.VisualStudio/Options/SarifViewerOptions.cs:
```diff
@@ -0,0 +1,80 @@
+        public static SarifViewerOption Instance
+        {
+            get;
+            private set;
```
>private [](start = 12, length = 7)

inline it, pls

yongyan-gh replied:
done

---
In reply to: [597772765](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r597772765) [](ancestors = 597772765)

eddynaka commented on src/Sarif.Viewer.VisualStudio/Options/SarifViewerOptionPage.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        private const string CategoryName = "Sarif Viewer Options";
+
+        private const string MonitorSarifFolderDisplayName = "Enable loading sarif results in .sarif folder automatically";
```
>in [](start = 91, length = 2)

from

yongyan-gh replied:
updated

---
In reply to: [598007772](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r598007772) [](ancestors = 598007772)

eddynaka commented on src/Sarif.Viewer.VisualStudio/Options/SarifViewerOptionPage.cs:
```diff
@@ -0,0 +1,25 @@
+namespace Microsoft.Sarif.Viewer.Options
+{
+    [ComVisible(true)]
+    public class SarifViewerOptionPage : DialogPage
```
>SarifViewerOptionPage [](start = 17, length = 21)

imagine that it started as true.
when u change to false, does it trigger any event to cancel previous watchers?

yongyan-gh replied:
Option value was checked in StartWatch/StopWatch methods, it only happens  when solution is loaded/closed.
If option changed, the change would effect next time solution is loaded/closed, before that any file added/deleted still be watched.
In latest change moved the checking to from Start/StopWatch to file created/deleted events, so the change will effect immediately. 

---
In reply to: [598008582](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r598008582) [](ancestors = 598008582)

eddynaka commented:
>                     fileSystemWatcher.Renamed -= this.SarifLogFile_Renamed;

it's missing two events: created and deleted

---
Refers to: src/Sarif.Viewer.VisualStudio/FileWatcher/FileWatcher.cs:85 in 87e2638. [](commit_id = 87e26388768ea53a1fb1d0e10b786c0114eadc5b, deletion_comment = False)

eddynaka commented on src/Sarif.Viewer.VisualStudio/FileWatcher/SarifFolderMonitor.cs:
```diff
@@ -0,0 +1,169 @@
+        public const string SarifFileExtensionName = ".sarif";
+
+        private readonly IFileSystem fileSystem;
+        private IFileWatcher fileWatcher;
```
>IFileWatcher fileWatcher [](start = 16, length = 24)

maybe you can add readonly

eddynaka commented on src/Sarif.Viewer.VisualStudio/FileWatcher/SarifFolderMonitor.cs:
```diff
@@ -0,0 +1,169 @@
+        private readonly IFileSystem fileSystem;
+        private IFileWatcher fileWatcher;
+        private bool filesLoaded;
+        private readonly ILoadSarifLogService sarifLoadLogService;
```
>private readonly [](start = 8, length = 16)

group all readonly first

yongyan-gh replied:
done

---
In reply to: [602304497](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r602304497) [](ancestors = 602304497)

eddynaka commented on src/Sarif.Viewer.VisualStudio/Options/SarifViewerOptionPage.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        private const string CategoryName = "Sarif Viewer Options";
+
+        private const string MonitorSarifFolderDisplayName = "Enable loading sarif results from .sarif folder automatically";
```
>sarif results [](start = 77, length = 13)

.sarif files

yongyan-gh replied:
fixed

---
In reply to: [602306723](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r602306723) [](ancestors = 602306723)

eddynaka commented on src/Sarif.Viewer.VisualStudio/FileWatcher/SarifFolderMonitor.cs:
```diff
@@ -0,0 +1,169 @@
+            this.fileWatcher.Filter = Constants.SarifFileSearchPattern;
+
+            // here no need to watch for sarif file log updates
+            // because when load the sarif log file in viewer, its already monitored by viewer's file watcher
```
// no need to watch for the sarif file log updates
// because when we load, the sarif log file in the viewer, it's already monitored by the viewer's file watcher

yongyan-gh replied:
corrected

---
In reply to: [602309039](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r602309039) [](ancestors = 602309039)

eddynaka commented on src/Sarif.Viewer.VisualStudio/FileWatcher/FileWatcher.cs:
```diff
@@ -96,8 +146,13 @@ private void SarifLogFile_Changed(object sender, FileSystemEventArgs e)
             try
             {
-                // to avoid trigger changed event multiple times in a short time period.
+                // some processes update files using complex mechanisim, may cause file changed
```
>mechanisim [](start = 61, length = 10)

mechanism

yongyan-gh replied:
corrected

---
In reply to: [602309794](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r602309794) [](ancestors = 602309794)

yongyan-gh commented:
>                     fileSystemWatcher.Renamed -= this.SarifLogFile_Renamed;

fixed

---
In reply to: [808229407](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#issuecomment-808229407) [](ancestors = 808229407)

---
Refers to: src/Sarif.Viewer.VisualStudio/FileWatcher/FileWatcher.cs:85 in 87e2638. [](commit_id = 87e26388768ea53a1fb1d0e10b786c0114eadc5b, deletion_comment = False)

eddynaka commented on src/Sarif.Viewer.VisualStudio/FileWatcher/SarifFolderMonitor.cs:
```diff
@@ -0,0 +1,183 @@
+                this.fileWatcher.Filter = Constants.SarifFileSearchPattern;
+
+                // no need to watch for the sarif file log updates
+                // because when we load, the sarif log file in the viewer, it's already monitored by the viewer's file watcher
```
>, [](start = 39, length = 1)

remove this , i think it was added by my mistake
